From 3fe994d7b852349b14bceab019d9b9e5e831813d Mon Sep 17 00:00:00 2001
From: Alisher Alikhodjaev <alisher@google.com>
Date: Tue, 24 Jan 2023 21:37:47 +0000
Subject: [PATCH] Revert "OOBR in NxpMfcReader::SendIncDecRestoreCmdPart2"

This reverts commit 6dfdec6cfa0f858a25b35beed6a928d5838c11af.

Reason for revert: CVE-2022-20471 re-released due to functional regression

Change-Id: I904041f179a50f56bd6454e3f414915c9ab61edd
Merged-In: Ie6ae772c695648fb6dc4ae80a25e5c4e287de3f2
(cherry picked from commit f61bd379e89790af74414ef49061da67e1910497)
Merged-In: I904041f179a50f56bd6454e3f414915c9ab61edd
---
 pn8x/halimpl/mifare/NxpMfcReader.cc  | 10 ++--------
 pn8x/halimpl/mifare/NxpMfcReader.h   |  4 ++--
 snxxx/halimpl/mifare/NxpMfcReader.cc | 10 ++--------
 snxxx/halimpl/mifare/NxpMfcReader.h  |  2 +-
 4 files changed, 7 insertions(+), 19 deletions(-)

diff --git a/pn8x/halimpl/mifare/NxpMfcReader.cc b/pn8x/halimpl/mifare/NxpMfcReader.cc
index 23a482e..ba850eb 100644
--- a/pn8x/halimpl/mifare/NxpMfcReader.cc
+++ b/pn8x/halimpl/mifare/NxpMfcReader.cc
@@ -59,7 +59,7 @@ int NxpMfcReader::Write(uint16_t mfcDataLen, const uint8_t* pMfcData) {
   /* send TAG_CMD part 2 for Mifare increment ,decrement and restore commands */
   if (mfcTagCmdBuff[4] == eMifareDec || mfcTagCmdBuff[4] == eMifareInc ||
       mfcTagCmdBuff[4] == eMifareRestore) {
-    SendIncDecRestoreCmdPart2(mfcDataLen, pMfcData);
+    SendIncDecRestoreCmdPart2(pMfcData);
   }
   return writtenDataLen;
 }
@@ -263,20 +263,14 @@ void NxpMfcReader::AuthForWrite() {
 ** Returns          None
 **
 *******************************************************************************/
-void NxpMfcReader::SendIncDecRestoreCmdPart2(uint16_t mfcDataLen,
-                                             const uint8_t* mfcData) {
+void NxpMfcReader::SendIncDecRestoreCmdPart2(const uint8_t* mfcData) {
   NFCSTATUS status = NFCSTATUS_SUCCESS;
   /* Build TAG_CMD part 2 for Mifare increment ,decrement and restore commands*/
   uint8_t incDecRestorePart2[] = {0x00, 0x00, 0x05, (uint8_t)eMfRawDataXchgHdr,
                                   0x00, 0x00, 0x00, 0x00};
   uint8_t incDecRestorePart2Size =
       (sizeof(incDecRestorePart2) / sizeof(incDecRestorePart2[0]));
-
   if (mfcData[3] == eMifareInc || mfcData[3] == eMifareDec) {
-    if (incDecRestorePart2Size >= mfcDataLen) {
-      incDecRestorePart2Size = mfcDataLen - 1;
-      android_errorWriteLog(0x534e4554, "238177877");
-    }
     for (int i = 4; i < incDecRestorePart2Size; i++) {
       incDecRestorePart2[i] = mfcData[i + 1];
     }
diff --git a/pn8x/halimpl/mifare/NxpMfcReader.h b/pn8x/halimpl/mifare/NxpMfcReader.h
index 97f6ccb..8820eaf 100644
--- a/pn8x/halimpl/mifare/NxpMfcReader.h
+++ b/pn8x/halimpl/mifare/NxpMfcReader.h
@@ -109,7 +109,7 @@ class NxpMfcReader {
   void BuildIncDecCmd();
   void CalcSectorAddress();
   void AuthForWrite();
-  void SendIncDecRestoreCmdPart2(uint16_t mfcDataLen, const uint8_t* mfcData);
+  void SendIncDecRestoreCmdPart2(const uint8_t* mfcData);
 
  public:
   int Write(uint16_t mfcDataLen, const uint8_t* pMfcData);
@@ -117,4 +117,4 @@ class NxpMfcReader {
   NFCSTATUS CheckMfcResponse(uint8_t* pTransceiveData,
                              uint16_t transceiveDataLen);
   static NxpMfcReader& getInstance();
-};
+};
\ No newline at end of file
diff --git a/snxxx/halimpl/mifare/NxpMfcReader.cc b/snxxx/halimpl/mifare/NxpMfcReader.cc
index 586d287..879953b 100644
--- a/snxxx/halimpl/mifare/NxpMfcReader.cc
+++ b/snxxx/halimpl/mifare/NxpMfcReader.cc
@@ -69,7 +69,7 @@ int NxpMfcReader::Write(uint16_t mfcDataLen, const uint8_t* pMfcData) {
     MfcWaitForAck();
     if (isAck) {
       NXPLOG_NCIHAL_D("part 1 command Acked");
-      SendIncDecRestoreCmdPart2(mfcDataLen, pMfcData);
+      SendIncDecRestoreCmdPart2(pMfcData);
     } else {
       NXPLOG_NCIHAL_E("part 1 command NACK");
     }
@@ -297,20 +297,14 @@ void NxpMfcReader::AuthForWrite() {
 ** Returns          None
 **
 *******************************************************************************/
-void NxpMfcReader::SendIncDecRestoreCmdPart2(uint16_t mfcDataLen,
-                                             const uint8_t* mfcData) {
+void NxpMfcReader::SendIncDecRestoreCmdPart2(const uint8_t* mfcData) {
   NFCSTATUS status = NFCSTATUS_SUCCESS;
   /* Build TAG_CMD part 2 for Mifare increment ,decrement and restore commands*/
   uint8_t incDecRestorePart2[] = {0x00, 0x00, 0x05, (uint8_t)eMfRawDataXchgHdr,
                                   0x00, 0x00, 0x00, 0x00};
   uint8_t incDecRestorePart2Size =
       (sizeof(incDecRestorePart2) / sizeof(incDecRestorePart2[0]));
-
   if (mfcData[3] == eMifareInc || mfcData[3] == eMifareDec) {
-    if (incDecRestorePart2Size >= mfcDataLen) {
-      incDecRestorePart2Size = mfcDataLen - 1;
-      android_errorWriteLog(0x534e4554, "238177877");
-    }
     for (int i = 4; i < incDecRestorePart2Size; i++) {
       incDecRestorePart2[i] = mfcData[i + 1];
     }
diff --git a/snxxx/halimpl/mifare/NxpMfcReader.h b/snxxx/halimpl/mifare/NxpMfcReader.h
index 7265fe3..f81c773 100644
--- a/snxxx/halimpl/mifare/NxpMfcReader.h
+++ b/snxxx/halimpl/mifare/NxpMfcReader.h
@@ -112,7 +112,7 @@ class NxpMfcReader {
   void BuildIncDecCmd();
   void CalcSectorAddress();
   void AuthForWrite();
-  void SendIncDecRestoreCmdPart2(uint16_t mfcDataLen, const uint8_t* mfcData);
+  void SendIncDecRestoreCmdPart2(const uint8_t* mfcData);
 
  public:
   int Write(uint16_t mfcDataLen, const uint8_t* pMfcData);
-- 
2.39.1.456.gfc5497dd1b-goog

